experimental/argset-chaining/chainer.metta

```diff
@@ -0,0 +1,162 @@
+(= (myfold-atom $list $init $f)
+   (if-decons-expr $list $head $tail
+      (myfold-atom $tail ($f $init $head) $f)
+      $init))
+
+
+;;Subtract one set from another
+(= ((sub_ff $list) $accum $elem)
+   (if (elem $list $elem) $accum (cons-atom $elem $accum)))
+
+(: substract (-> Expression Expression Expression))
+(= (substract $list1 $list2)
+   (myfold-atom $list1 () (sub_ff $list2)))
+
+;;Remove one element from a set
+(= (remove $list $elem)
+   (if-decons-expr $list $head $tail
+      (if (== $head $elem) $tail (let $new_tail (remove $tail $elem) (cons-atom $head $new_tail)))
+      $list))
+
+;;Check if two sets are equal
+(= (seteq_ff ($b $lst) $elem)
+     (if (elem $lst $elem) ($b (remove $lst $elem)) (False $lst)))
+
+(= (seteq $list1 $list2)
+   (if (or (== (get-metatype $list1) Variable) (== (get-metatype $list2) Variable)) True
+   (let ($b $l2) (myfold-atom $list1 (True $list2) seteq_ff)
+     (if (or (not $b) (not (== $l2 ()))) False True)
+   )))
+
+
+;;Check if one set is a subset of another
+(= (subset $list1 $list2)
+   (if (or (== (get-metatype $list1) Variable) (== (get-metatype $list2) Variable)) True
+   (let ($b $l2) (myfold-atom $list1 (True $list2) seteq_ff) $b)))
+
+
+;;Concat 2 sets together
+(= (concat $list1 $list2)
+   (if (== (get-metatype $list1) Variabel)
+     $list2
+     (if (== (get-metatype $list2) Variable)
+     $list1
+     (union-atom $list1 $list2))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; To simplify the chainer we assume that everything is a function
+;; So (: a A) needs to be exppressed as (: (a ()) (-> () A))
+
+;; To convert to the function from we have the following
+(= (cleanThrm $thrm) (unify $thrm (-> $prms $ccls) (-> $prms $ccls) (-> () $thrm)))
+(= (cleanPrf $prf) (unify $prf ($abs $args) ($abs $args) ($prf ())))
+(= (cleanStmt (: $prf $thrm)) (: (cleanPrf $prf) (cleanThrm $thrm)))
+
+;; And this is to convert back to the original form
+(= (prettyThrm $thrm) (unify $thrm (-> () $ccls) $ccls $thrm))
+(= (prettyPrf $prf) (unify $prf ($abs ()) $abs $prf))
+(= (prettyStmt (: $prf $thrm)) (: (prettyPrf $prf) (prettyThrm $thrm)))
+
+!(assertEqualToResult (prettyStmt (cleanStmt (: a A))) ((: a A)))
+
+;;;;;;;;;;;;;;;;;;;;;;
+;; Backward chainer ;;
+;;;;;;;;;;;;;;;;;;;;;;
+(: bc (-> $a                            ; Knowledge base space
+          Nat                           ; Maximum depth
+          $q                            ; Query
+          $r))                          ; Result
+
+(= (bc $kb $depth_full $query) 
+ ;Base Case
+ (unify $kb $query $query
+  ;Recursive steps
+  (let* (((S $depth) $depth_full)
+         ((: ($abs $args) (-> $prms $conclusion)) (cleanStmt $query))
+         ((: $matched_prf (-> $matched_prms $conclusion))
+          (bc $kb $depth (: $matched_prf (-> $matched_prms $conclusion))))
+
+         (($matched_abs $matched_args) (cleanPrf $matched_prf))
+
+         (() (unify $abs $matched_abs () Empty))
```

Comment:
Can't you replace that by
```suggestion
         ($abs $matched_abs)
```
since it is part of `let*`.

Reply:
Yea i just realized that myself.